RegisterHostname causes panic if 2 filebeat receivers are configured

If 2 filebeat receivers are configured, and run a panic occurs when the second tries to register it's hostname in the metrics.


https://github.com/elastic/beats/blob/566e63b3372755800e4091d8331afe49edcab630/libbeat/cmd/instance/beat.go#L839

This is because the monitoring package uses a package level global for the monitoring namespace.

https://github.com/elastic/elastic-agent-libs/blob/4a2d1a91a0431a3adaab859025d8b19baeddd5a3/monitoring/namespace.go#L22

We need to figure out how to have unique monitoring per beat receiver

```
elastic2:elastic-agent-9.0.0-darwin-aarch64% ./elastic-agent otel -c otel.yml
Starting in otel mode
2024-11-21T20:10:43.289-0600	info	service@v0.113.0/service.go:166	Setting up own telemetry...
2024-11-21T20:10:43.289-0600	info	telemetry/metrics.go:70	Serving metrics	{"address": "localhost:8888", "metrics level": "Normal"}
2024-11-21T20:10:43.295-0600	info	Home path: [] Config path: [] Data path: [data] Logs path: [logs]	{"kind": "receiver", "name": "filebeatreceiver/2", "data_type": "logs", "service.name": "elastic2", "ecs.version": "1.6.0"}
2024-11-21T20:10:43.307-0600	info	Beat ID: f5d4082a-3dd7-47cb-8e5d-9caff65644e7	{"kind": "receiver", "name": "filebeatreceiver/2", "data_type": "logs", "service.name": "elastic2", "ecs.version": "1.6.0"}
2024-11-21T20:10:43.339-0600	warn	unable to lookup FQDN: could not get FQDN, all methods failed: failed looking up CNAME: lookup elastic2 on 172.16.1.1:53: no such host: failed looking up IP: lookup elastic2: no such host, using hostname = elastic2 as FQDN	{"kind": "receiver", "name": "filebeatreceiver/2", "data_type": "logs", "service.name": "elastic2", "ecs.version": "1.6.0"}
2024-11-21T20:10:43.340-0600	info	publisher	Beat name: elastic2	{"kind": "receiver", "name": "filebeatreceiver/2", "data_type": "logs", "service.name": "elastic2", "ecs.version": "1.6.0"}
2024-11-21T20:10:43.340-0600	error	modules	Not loading modules. Module directory not found: module	{"kind": "receiver", "name": "filebeatreceiver/2", "data_type": "logs", "service.name": "elastic2", "ecs.version": "1.6.0"}
2024-11-21T20:10:43.340-0600	warn	Filebeat is unable to load the ingest pipelines for the configured modules because the Elasticsearch output is not configured/enabled. If you have already loaded the ingest pipelines or are using Logstash pipelines, you can ignore this warning.	{"kind": "receiver", "name": "filebeatreceiver/2", "data_type": "logs", "service.name": "elastic2", "ecs.version": "1.6.0"}
panic: the variable host cannot be re-registered with options

goroutine 1 [running]:
github.com/elastic/elastic-agent-libs/monitoring.panicErr(...)
	github.com/elastic/elastic-agent-libs@v0.17.3/monitoring/registry.go:257
github.com/elastic/elastic-agent-libs/monitoring.setupMetric(0x1400238c880, {0x1086f639e, 0x4}, {0x14003667610?, 0x1, 0x111460108?})
	github.com/elastic/elastic-agent-libs@v0.17.3/monitoring/metrics.go:350 +0xfc
github.com/elastic/elastic-agent-libs/monitoring.NewFunc(0x1400237ff00?, {0x1086f639e, 0x4}, 0x14002382810, {0x14003667610, 0x1, 0x1})
	github.com/elastic/elastic-agent-libs@v0.17.3/monitoring/metrics.go:242 +0x44
github.com/elastic/beats/v7/libbeat/cmd/instance.(*Beat).RegisterHostname(0x14003112870?, 0x8?)
	github.com/elastic/beats/v7@v7.0.0-alpha2.0.20241111113032-56dca4bbda40/libbeat/cmd/instance/beat.go:839 +0x144
github.com/elastic/beats/v7/libbeat/cmd/instance.NewBeatReceiver({{0x108705204, 0x8}, {0x0, 0x0}, {0x0, 0x0}, 0x1, 0x1, {{0x0, 0x0}, ...}, ...}, ...)
	github.com/elastic/beats/v7@v7.0.0-alpha2.0.20241111113032-56dca4bbda40/libbeat/cmd/instance/beat.go:378 +0x9a4
github.com/elastic/beats/v7/x-pack/filebeat/fbreceiver.createReceiver({0x10339adac?, 0x3?}, {{{{0x14001ee0f00, 0x10}}, {0x14001ee0f11, 0x1}}, {0x1400227b180, {0x10b493f00, 0x14001d20240}, {0x10b48ca88, ...}, ...}, ...}, ...)
	github.com/elastic/beats/v7@v7.0.0-alpha2.0.20241111113032-56dca4bbda40/x-pack/filebeat/fbreceiver/factory.go:45 +0x434
go.opentelemetry.io/collector/receiver.CreateLogsFunc.CreateLogs(...)
	go.opentelemetry.io/collector/receiver@v0.113.0/receiver.go:133
go.opentelemetry.io/collector/service/internal/builders.(*ReceiverBuilder).CreateLogs(0x1400237de40, {0x10b4f0b28, 0x1400228a370}, {{{{0x14001ee0f00, 0x10}}, {0x14001ee0f11, 0x1}}, {0x1400227b180, {0x10b493f00, 0x14001d20240}, ...}, ...}, ...)
	go.opentelemetry.io/collector/service@v0.113.0/internal/builders/receiver.go:86 +0x10c
go.opentelemetry.io/collector/service/internal/graph.(*receiverNode).buildComponent(0x1400228bd60, {0x10b4f0b28, 0x1400228a370}, {0x1400227b180, {0x10b493f00, 0x14001d20240}, {0x10b48ca88, 0x1400237b7a0}, 0x1400237f500, 0x1, ...}, ...)
	go.opentelemetry.io/collector/service@v0.113.0/internal/graph/receiver.go:67 +0x6fc
go.opentelemetry.io/collector/service/internal/graph.(*Graph).buildComponents(0x14000171c80, {0x10b4f0b28, 0x1400228a370}, {{0x14002339000, {0x10b493f00, 0x14001d20240}, {0x10b48ca88, 0x1400237b7a0}, 0x1400237f500, 0x1, ...}, ...})
	go.opentelemetry.io/collector/service@v0.113.0/internal/graph/graph.go:291 +0x618
go.opentelemetry.io/collector/service/internal/graph.Build({0x10b4f0b28, 0x1400228a370}, {{0x14002339000, {0x10b493f00, 0x14001d20240}, {0x10b48ca88, 0x1400237b7a0}, 0x1400237f500, 0x1, {0x1400237f0c0, ...}}, ...})
	go.opentelemetry.io/collector/service@v0.113.0/internal/graph/graph.go:90 +0x348
go.opentelemetry.io/collector/service.(*Service).initGraph(_, {_, _}, {{{0x0, 0x0, {0x108700951, 0x7}, 0x0, 0x0, 0x1400206b620, ...}, ...}, ...})
	go.opentelemetry.io/collector/service@v0.113.0/service.go:342 +0x150
go.opentelemetry.io/collector/service.New({_, _}, {{{0x16d8b77a0, 0xf}, {0x10882db41, 0x2c}, {0x14000f83fd8, 0x5}}, 0x14002354fe0, 0x1400232d200, ...}, ...)
	go.opentelemetry.io/collector/service@v0.113.0/service.go:195 +0xca8
go.opentelemetry.io/collector/otelcol.(*Collector).setupConfigurationComponents(0x14002284a20, {0x10b4f0b28, 0x1400228a370})
	go.opentelemetry.io/collector/otelcol@v0.113.0/collector.go:183 +0x414
go.opentelemetry.io/collector/otelcol.(*Collector).Run(0x14002284a20, {0x10b4f0b28, 0x1400228a370})
	go.opentelemetry.io/collector/otelcol@v0.113.0/collector.go:282 +0x40
github.com/elastic/elastic-agent/internal/pkg/otel.Run({0x10b4f0b28, 0x1400228a0f0}, 0x140010ba360, {0x140010b0bc0, 0x1, 0x1})
	github.com/elastic/elastic-agent/internal/pkg/otel/run.go:44 +0x190
github.com/elastic/elastic-agent/internal/pkg/agent/cmd.runCollector({0x10b4f0a80, 0x1102ece88}, {0x140010b0bc0, 0x1, 0x1})
	github.com/elastic/elastic-agent/internal/pkg/agent/cmd/otel.go:81 +0x148
github.com/elastic/elastic-agent/internal/pkg/agent/cmd.newOtelCommandWithArgs.func1(0x140022cc008, {0x1086f631a?, 0x4?, 0x1086f61f6?})
	github.com/elastic/elastic-agent/internal/pkg/agent/cmd/otel.go:32 +0x64
github.com/spf13/cobra.(*Command).execute(0x140022cc008, {0x14002286840, 0x2, 0x2})
	github.com/spf13/cobra@v1.8.1/command.go:985 +0x840
github.com/spf13/cobra.(*Command).ExecuteC(0x1400027cc08)
	github.com/spf13/cobra@v1.8.1/command.go:1117 +0x344
github.com/spf13/cobra.(*Command).Execute(0x140000b2080?)
	github.com/spf13/cobra@v1.8.1/command.go:1041 +0x1c
main.main()
	github.com/elastic/elastic-agent/main.go:32 +0x130
```

--- a/filebeat/beater/filebeat.go
+++ b/filebeat/beater/filebeat.go
@@ -272,10 +272,18 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
 	waitEvents := newSignalWait()
 
 	// count active events for waiting on shutdown
+	var reg *monitoring.Registry
+
+	if b.Info.Monitoring.Namespace != nil {
+		reg = b.Info.Monitoring.Namespace.GetRegistry().GetRegistry("stats")
+		if reg == nil {
+			reg = b.Info.Monitoring.Namespace.GetRegistry().NewRegistry("stats")
+		}
+	}
 	wgEvents := &eventCounter{
-		count: monitoring.NewInt(nil, "filebeat.events.active"), // Gauge
-		added: monitoring.NewUint(nil, "filebeat.events.added"),
-		done:  monitoring.NewUint(nil, "filebeat.events.done"),
+		count: monitoring.NewInt(reg, "filebeat.events.active"), // Gauge
+		added: monitoring.NewUint(reg, "filebeat.events.added"),
+		done:  monitoring.NewUint(reg, "filebeat.events.done"),
 	}
 	finishedLogger := newFinishedLogger(wgEvents)
 

--- a/heartbeat/monitors/mocks.go
+++ b/heartbeat/monitors/mocks.go
@@ -60,12 +60,8 @@ func makeMockFactory(pluginsReg *plugin.PluginsReg) (factory *RunnerFactory, sch
 		EphemeralID:     eid,
 		FirstStart:      time.Now(),
 		StartTime:       time.Now(),
-		Monitoring: struct {
-			DefaultUsername string
-		}{
-			DefaultUsername: "test",
-		},
 	}
+	info.Monitoring.DefaultUsername = "test"
 
 	sched = scheduler.Create(
 		1,
@@ -246,7 +242,8 @@ func mockPluginBuilder() (plugin.PluginFactory, *atomic.Int, *atomic.Int) {
 
 				return plugin.Plugin{Jobs: j, DoClose: closer, Endpoints: 1}, nil
 			},
-			Stats: plugin.NewPluginCountersRecorder("test", reg)},
+			Stats: plugin.NewPluginCountersRecorder("test", reg),
+		},
 		built,
 		closed
 }

--- a/libbeat/beat/info.go
+++ b/libbeat/beat/info.go
@@ -22,6 +22,8 @@ import (
 
 	"github.com/gofrs/uuid/v5"
 	"go.opentelemetry.io/collector/consumer"
+
+	"github.com/elastic/elastic-agent-libs/monitoring"
 )
 
 // Info stores a beats instance meta data.
@@ -41,9 +43,10 @@ type Info struct {
 
 	// Monitoring-related fields
 	Monitoring struct {
-		DefaultUsername string // The default username to be used to connect to Elasticsearch Monitoring
+		DefaultUsername string                // The default username to be used to connect to Elasticsearch Monitoring
+		Namespace       *monitoring.Namespace // a monitor namespace that is unique per beat instance
 	}
-	LogConsumer consumer.Logs //otel log consumer
+	LogConsumer consumer.Logs // otel log consumer
 
 }
 

--- a/libbeat/cmd/instance/beat.go
+++ b/libbeat/cmd/instance/beat.go
@@ -337,6 +337,8 @@ func NewBeatReceiver(settings Settings, receiverConfig map[string]interface{}, c
 		config.OverwriteConfigOpts(configOpts(store))
 	}
 
+	b.Beat.Info.Monitoring.Namespace = monitoring.GetNamespace(b.Info.Beat + "-" + b.Info.ID.String())
+
 	instrumentation, err := instrumentation.New(cfg, b.Info.Beat, b.Info.Version)
 	if err != nil {
 		return nil, fmt.Errorf("error setting up instrumentation: %w", err)
@@ -469,11 +471,6 @@ func NewBeatReceiver(settings Settings, receiverConfig map[string]interface{}, c
 		return nil, fmt.Errorf("error creating processors: %w", err)
 	}
 
-	reg := monitoring.Default.GetRegistry(b.Info.Name)
-	if reg == nil {
-		reg = monitoring.Default.NewRegistry(b.Info.Name)
-	}
-
 	// This should be replaced with static config for otel consumer
 	// but need to figure out if we want the Queue settings from here.
 	outputEnabled := b.Config.Output.IsSet() && b.Config.Output.Config().Enabled()
@@ -485,12 +482,14 @@ func NewBeatReceiver(settings Settings, receiverConfig map[string]interface{}, c
 		}
 	}
 
-	tel := reg.GetRegistry("state")
+	uniq_reg := b.Beat.Info.Monitoring.Namespace.GetRegistry()
+
+	tel := uniq_reg.GetRegistry("state")
 	if tel == nil {
-		tel = reg.NewRegistry("state")
+		tel = uniq_reg.NewRegistry("state")
 	}
 	monitors := pipeline.Monitors{
-		Metrics:   reg,
+		Metrics:   uniq_reg,
 		Telemetry: tel,
 		Logger:    logp.L().Named("publisher"),
 		Tracer:    b.Instrumentation.Tracer(),
@@ -510,7 +509,6 @@ func NewBeatReceiver(settings Settings, receiverConfig map[string]interface{}, c
 	b.Publisher = publisher
 
 	return b, nil
-
 }
 
 // InitWithSettings does initialization of things common to all actions (read confs, flags)
@@ -831,11 +829,27 @@ func (b *Beat) RegisterHostname(useFQDN bool) {
 	hostname := b.Info.FQDNAwareHostname(useFQDN)
 
 	// info.hostname
-	infoRegistry := monitoring.GetNamespace("info").GetRegistry()
+	var infoRegistry *monitoring.Registry
+	if b.Info.Monitoring.Namespace != nil {
+		infoRegistry = b.Info.Monitoring.Namespace.GetRegistry().GetRegistry("info")
+		if infoRegistry == nil {
+			infoRegistry = b.Info.Monitoring.Namespace.GetRegistry().NewRegistry("info")
+		}
+	} else {
+		infoRegistry = monitoring.GetNamespace("info").GetRegistry()
+	}
 	monitoring.NewString(infoRegistry, "hostname").Set(hostname)
 
 	// state.host
-	stateRegistry := monitoring.GetNamespace("state").GetRegistry()
+	var stateRegistry *monitoring.Registry
+	if b.Info.Monitoring.Namespace != nil {
+		stateRegistry = b.Info.Monitoring.Namespace.GetRegistry().GetRegistry("state")
+		if stateRegistry == nil {
+			stateRegistry = b.Info.Monitoring.Namespace.GetRegistry().NewRegistry("state")
+		}
+	} else {
+		stateRegistry = monitoring.GetNamespace("state").GetRegistry()
+	}
 	monitoring.NewFunc(stateRegistry, "host", host.ReportInfo(hostname), monitoring.Report)
 }
 

--- a/x-pack/filebeat/input/benchmark/input.go
+++ b/x-pack/filebeat/input/benchmark/input.go
@@ -60,7 +60,7 @@ func (bi *benchmarkInput) Test(ctx v2.TestContext) error {
 // Run starts the data generation.
 func (bi *benchmarkInput) Run(ctx v2.Context, publisher stateless.Publisher) error {
 	var wg sync.WaitGroup
-	metrics := newInputMetrics(ctx.ID)
+	metrics := newInputMetrics(ctx)
 
 	for i := uint8(0); i < bi.cfg.Threads; i++ {
 		wg.Add(1)
@@ -103,8 +103,8 @@ func runThread(ctx v2.Context, publisher stateless.Publisher, thread uint8, cfg 
 				ticker.Stop()
 				return
 			case <-ticker.C:
-				//don't want to block on filling doPublish channel
-				//so only send as many as it can hold right now
+				// don't want to block on filling doPublish channel
+				// so only send as many as it can hold right now
 				numToSend := cap(pubChan) - len(pubChan)
 				for i := 0; i < numToSend; i++ {
 					pubChan <- true
@@ -157,8 +157,8 @@ type inputMetrics struct {
 }
 
 // newInputMetrics returns an input metric for the benchmark processor.
-func newInputMetrics(id string) *inputMetrics {
-	reg, unreg := inputmon.NewInputRegistry(inputName, id, nil)
+func newInputMetrics(ctx v2.Context) *inputMetrics {
+	reg, unreg := inputmon.NewInputRegistry(inputName, ctx.ID, ctx.Agent.Monitoring.Namespace.GetRegistry())
 	out := &inputMetrics{
 		unregister:      unreg,
 		eventsPublished: monitoring.NewUint(reg, "events_published_total"),

--- a/x-pack/heartbeat/scenarios/framework/framework.go
+++ b/x-pack/heartbeat/scenarios/framework/framework.go
@@ -30,11 +30,13 @@ import (
 	beatversion "github.com/elastic/beats/v7/libbeat/version"
 )
 
-type ScenarioRun func(t *testing.T) (config mapstr.M, meta ScenarioRunMeta, close func(), err error)
-type ScenarioRunMeta struct {
-	URL    *url.URL
-	Status monitorstate.StateStatus
-}
+type (
+	ScenarioRun     func(t *testing.T) (config mapstr.M, meta ScenarioRunMeta, close func(), err error)
+	ScenarioRunMeta struct {
+		URL    *url.URL
+		Status monitorstate.StateStatus
+	}
+)
 
 type Scenario struct {
 	Name         string
@@ -155,7 +157,6 @@ func NewScenarioDB() *ScenarioDB {
 		ByTag:    map[string][]Scenario{},
 		All:      []Scenario{},
 	}
-
 }
 
 func (sdb *ScenarioDB) Init() {
@@ -250,7 +251,9 @@ func runMonitorOnce(t *testing.T, monitorConfig mapstr.M, meta ScenarioRunMeta, 
 
 	mIface, err := f.Create(pipe, conf)
 	require.NoError(t, err)
-	mtr.monitor = mIface.(*monitors.Monitor)
+	mon, ok := mIface.(*monitors.Monitor)
+	require.True(t, ok, "type assertion didn't succeed")
+	mtr.monitor = mon
 	require.NotNil(t, mtr.monitor, "could not convert to monitor %v", mIface)
 	mtr.Events = pipe.PublishedEvents
 
@@ -281,12 +284,8 @@ func setupFactoryAndSched(location *hbconfig.LocationWithID, stateLoader monitor
 		EphemeralID:     eid,
 		FirstStart:      time.Now(),
 		StartTime:       time.Now(),
-		Monitoring: struct {
-			DefaultUsername string
-		}{
-			DefaultUsername: "test",
-		},
 	}
+	info.Monitoring.DefaultUsername = "test"
 
 	sched = scheduler.Create(
 		1,

